spec/fastsync/version2/scheduler.tla

```diff
@@ -0,0 +1,580 @@
+       /\ outEvent' = [type |-> "scPeerError", peerID |-> inEvent.peerID, reason |-> res.err]
+       /\ scheduler' = res1.val
+     ELSE
+       /\ outEvent' = [type |-> "scBlockReceived", peerID |-> inEvent.peerID, block |-> inEvent.block]
+       /\ scheduler' = res.val
+  /\ UNCHANGED scRunning
+
+handleNoBlockResponse ==
+  /\ inEvent.type = "bcNoBlockResponse"
+  /\ IF (scheduler.peers = {} \/ scheduler.peerStates[inEvent.peerID] = "peerStateRemoved") THEN
+        /\ outEvent' = noEvent
+        /\ UNCHANGED scheduler
+     ELSE
+       LET res == removePeer(scheduler, inEvent.peerID) IN
+        /\ outEvent' = [type |-> "scPeerError", peerID |-> inEvent.peerID, reason |-> "errPeerNoBlock"]
+        /\ scheduler' = res.val
+  /\ UNCHANGED scRunning
+
+handleBlockProcessed ==
+  /\ inEvent.type = "pcBlockProcessed"
+  /\ IF inEvent.height /= scheduler.height THEN
+       /\ outEvent' = [type |-> "scSchedulerFail", reason |-> "errProcessedBlockEv"]
+       /\ UNCHANGED scheduler
+     ELSE
+       LET res == markProcessed(scheduler, inEvent.height) IN
+       IF res.err /= noErr THEN
+         /\ outEvent' = [type |-> "scSchedulerFail", reason |-> res.err]
+         /\ UNCHANGED scheduler
+       ELSE
+         /\ scheduler' = res.val
+         /\ IF allBlocksProcessed(scheduler') THEN
+              outEvent' = [type |-> "scFinishedEv", reason |-> "finished"]
+            ELSE
+              outEvent' = noEvent
+  /\ UNCHANGED scRunning
+
+handleBlockProcessError ==
+  /\ inEvent.type = "pcBlockVerificationFailure"
+  /\ IF scheduler.peers = {} THEN
+        /\ outEvent' = noEvent
+        /\ UNCHANGED scheduler
+     ELSE
+        LET res1 == removePeer(scheduler, inEvent.firstPeerID) IN
+        LET res2 == removePeer(res1.val, inEvent.secondPeerID) IN
+          /\ IF allBlocksProcessed(res2.val) THEN
+               outEvent' = [type |-> "scFinishedEv", reason |-> "finished"]
+             ELSE
+               outEvent' = noEvent
+          /\ scheduler' = res2.val
+  /\ UNCHANGED scRunning
+
+onNoAdvanceExp ==
+    /\ inEvent.type = "tNoAdvanceExp"
+    /\ scRunning' = FALSE
+    /\ outEvent' = [type |-> "scFinishedEv", reason |-> "timeout"]
+    /\ UNCHANGED <<scheduler>>
+
+NextSc ==
+  IF ~scRunning
+    THEN outEvent' = noEvent /\ UNCHANGED <<scRunning, scheduler>>
+  ELSE
+    \/ handleStatusResponse
+    \/ handleAddNewPeer
+    \/ handleRemovePeer
+    \/ handleTrySchedule
+    \/ handleBlockResponse
+    \/ handleNoBlockResponse
+    \/ handleBlockProcessed
+    \/ handleBlockProcessError
+    \/ onNoAdvanceExp
+
+(* ----------------------------------------------------------------------------------------------*)
+(* The behavior of the environment.                                                              *)
+(* ----------------------------------------------------------------------------------------------*)
+
+InitEnv ==
+  /\ inEvent = noEvent
+  /\ envRunning = TRUE
```

Comment:
It is maybe not necessarily to have running flag for environment.It can always be running as we will effectively stop making state transitions after scheduler stops.

Reply:
i can't remember now but i had some trouble before without it. will try again.